### DISCUSSION TITLE
Frigate, PTZ cameras, and auto tracking

### DISCUSSION THREAD

**Discussion by hawkeye217** [blakeblackshear/frigate#6695] [Ideas]
I recently picked up a small Dahua (EmpireTech) PTZ camera with their latest auto tracking firmware on it. It's pretty slick. The built-in AI identifies humans and vehicles, and you can set up "zones" (Dahua calls them "IVS rules") that will trigger the PTZ movement of the camera to track the person or vehicle in real time. 

I ended up returning the camera because the lens wasn't ideal for the location I wanted it in. But it got me thinking... Since Frigate already does tracking perfectly (it was better than the Dahua, at least in the few days I was messing with it), it might not be that difficult of a task to write some code to turn any PTZ camera into a "poor man's autotracking" camera and keep a tracked object in the center of the frame.

There was definitely some smoothing, acceleration/deceleration, and interpolation happening on the Dahua as it tracked humans and cars. With the speed of Frigate's tracking and maybe a little bit of movement prediction, it might be possible to achieve with any old PTZ that can be controlled via onvif.

Has anyone thought of this before? Of course a lot of Frigate's code assumes a stationary camera, but perhaps parts of it could be leveraged for tracking and PTZ movement. Or maybe this is too far fetched and there are other limiting factors I'm not considering? 

Of course zones and other stationary camera features that Frigate offers would not be applicable with a moving PTZ camera. So maybe that could be specified as:

```
cameras:
  driveway:
    type: ptz <---- setting a camera type 
    ffmpeg:
      inputs: rtsp://
...
```

Just kinda thinking out loud. Would love any feedback!

**Comment by NickM-27** [comment]
Frigate doesn't do tracking perfectly. It has been a common issue and Blake recently committed a better tracker which we have found some additional improvements to be made as well. 

This new tracker does have support / features that address this scenario but there's a lot that goes into it and would need to be accounted for.

**Reply by hawkeye217** [reply]
> Frigate doesn't do tracking perfectly. It has been a common issue and Blake recently committed a better tracker which we have found some additional improvements to be made as well.
> 
> This new tracker does have support / features that address this scenario but there's a lot that goes into it and would need to be accounted for.

Maybe "perfectly" was the wrong word... "Better than the Dahua in my experience" would be the better phrase :)

Once my newer PTZ arrives, I may take a stab at some proof-of-concept code to see if it's even feasible.

**Reply by NickM-27** [reply]
Yeah, another consideration is tune-ability since different PTZ cameras will have different reaction times

**Reply by hawkeye217** [reply]
Right. I thought that some kind of initial calibration would be applicable to determine motor speed (pixels per second?) so that we could specify that in the config and ensure that any kind of panning or zooming would not take place if the object was moving too fast. 

You're totally right, though. Definitely lots to consider.

**Reply by danruck** [reply]
I've solved this by creating a Frigate zone on the left and right side of the frame.  I then set up (in the camera) 10 preset positions for the range of motion I want to move, eg: All the way left is 1, middle is 5, and all the way right is 10, with increments in between.  

I set up a number helper in Home assistant with a max value of 10, then set up an automation to increment it when a human enters the right zone and decrement it when a human enters the left zone.  Then I simply use another automation to send the ONVIF command to call the preset, based on the number in the helper.

It's a little complex to set up, but this has worked really well for ages and has the bonus of being able to reset to a chosen home position once the object has been gone for a few minutes.  I've used the same method with other NVRs and image recognition systems too (such as MotionEye and DOODS).

**Reply by hawkeye217** [reply]
That's a really great solution!

Do you have anything set up for tilting and zooming, or does your camera just pan when the human moves across the zones?

**Reply by danruck** [reply]
Thanks, I'm quite proud of it.  Tilting is taken care of in the preset points.  When looking down my driveway it's almost level, but when it's focused on my front door it tilts down a bit to get a better view.  My camera doesn't have zoom, but I imagine the same method would work for one that does.

**Reply by hawkeye217** [reply]
> Frigate doesn't do tracking perfectly. It has been a common issue and Blake recently committed a better tracker which we have found some additional improvements to be made as well.
> 
> This new tracker does have support / features that address this scenario but there's a lot that goes into it and would need to be accounted for.

I've been messing around with this a bit and wanted to ask your opinion. What do you think would be a good approach to take for a default behavior for multiple objects in the frame? If two people are detected, should the PTZ zoom into the area so that both of them are in the frame? Should we only focus on the closest/largest one?

I suppose these could be configurable behaviors, but I'm just looking for opinions on what might be the most normal or the most expected or helpful.

**Reply by NickM-27** [reply]
I have no idea to be honest, I'd be curious what the dahua / hikvision cameras that implement this feature do.

**Reply by hawkeye217** [reply]
https://github.com/blakeblackshear/frigate/assets/32435876/d952a36b-3f31-49bc-b514-ba6dae51a443

Here's an example from a Dahua cam that does autotracking.

**Reply by NickM-27** [reply]
That's pretty cool. As far as how it could work for frigate, maybe it would stay at a full overview until one of the objects enters a zone and then it would follow that object. If they both entered at the same time maybe it doesn't follow until one is larger (closer) than another

**Reply by hawkeye217** [reply]
https://github.com/blakeblackshear/frigate/assets/32435876/56a81fce-e40c-444b-9637-2d8ed890fffa

Here's another one from a Dahua PTZ. Interesting that it locks onto a single person and completely ignores the bikers riding by. A user on a Dahua forum said "My experience with 4 PTZs is that they will track the first trigger object until they lose track or are called away by a spotter cam or a time limit. If in a group, the PTZ is still only tracking one target, if the group happens to be close enough together, and the track target size is set relatively small you will see more of the group as mentioned above. But its still tracking a single specific target."

Frigate has the potential to be way better than this...

**Comment by hawkeye217** [comment]
Also, on a related note, I think the changes that @blakeblackshear is making to motion detection and region selection for 0.13 [here](https://github.com/blakeblackshear/frigate/pull/6741/) will be helpful for this use case!

**Reply by hawkeye217** [reply]
I wrote up a small proof-of-concept in Python to listen to MQTT messages from a dev container of Frigate and move a PTZ camera via onvif. It's working, but not without a few issues.

A few problems/issues came up:

- I get "ghost" detections as the PTZ moves, and of course this is expected. I'm seeing motion boxes in various spots around the scene, which varies slightly by adjusting the new ```lightning_threshold``` parameter, but without much effect to reducing the ghosts. Would it make sense to disable motion detection entirely as the PTZ moves and then reenable it after movement? Perhaps something else?
- Frigate reacquires the "person" once movement is complete, but the event ID is different (again, as expected). How can we ensure we're tracking the same object after the camera moves?

Would love any thoughts or ideas @NickM-27 @blakeblackshear

**Reply by hawkeye217** [reply]
A quick update - a lot of the ghosting issues are mitigated just by making sure Frigate sees the person as a true positive, ie: 

```payload['after']['false_positive'] == False```

Simple enough :)

**Reply by NickM-27** [reply]
I think it may be another change, MQTT messages are not sent until the object is no longer a false positive. Only the very first `before` message has false_positive = True. Perhaps you were reading that message though and it was causing issues

**Reply by hawkeye217** [reply]
Right.

This MQTT method is really just a proof-of-concept. I want to try it on a few other PTZs before even trying to modify Frigate itself. While there is certainly room for improvement, I'm not too displeased with the results so far.

**Comment by hawkeye217** [comment]
I created a PR for this feature here: https://github.com/blakeblackshear/frigate/pull/6913

It does basic panning and tilting, keeping the tracked object in the center of the field of view. Zooming will take a bit more thinking. But this should be a good start.
